### PR TITLE
Add the Image processing in Haskell with JuicyPixels and Repa tutorial

### PR DESCRIPTION
**Pull request by mrkkrp**
Here is what I have so far, it's more or less complete.

### CODE REVIEW

File: tutorials/haskell/image-processing/tutorial.md [hunk 1]
```diff
@@ -0,0 +1,737 @@
...
+      -- Internally pixel data is stored as consecutively packed
+      -- lines from top to bottom, scanned from left to right
+      -- within individual lines, from first to last color
+      -- component within each pixel.
+    , imageData   :: V.Vector (PixelBaseComponent a)
+    }
+```
+
+The definition is simple and intuitive, here we have width and height of
+image and single vector that contains pixel data.
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md:109]: @mrkkrp Can you rephrase this paragraph? I think changing the comma after intuitive for a colon or a dot would be better. Or maybe either removing the "here" or adding a comma after "here".

Something like: "The definition is simple and intuitive. We have the width and height of the image, and a single vector that contains pixel data."

File: tutorials/haskell/image-processing/tutorial.md [hunk 2]
```diff
@@ -0,0 +1,737 @@
...
+
+`{-# UNPACK #-}` is a different story. To make things lazy, Haskell by
+default represents everything as “boxed” data, that is, it stores a pointer
+to `Int` in `MyData`, not the `Int` itself:
+
+![Boxed data in Haskell](/tutorials/haskell/image-processing/boxed-data.png)
+
+[Thunk](https://wiki.haskell.org/Thunk) is name for data that not yet
+evaluated. When `{-# UNPACK #-}` pragma is used, data becomes part of parent
+structure:
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: @mrkkrp How about: "Thunk is a name for data that hasn't been evaluated yet. When the `{-# UNPACK #-}` pragma is used, the data becomes part of the the parent structure."?

File: tutorials/haskell/image-processing/tutorial.md [hunk 3]
```diff
@@ -0,0 +1,737 @@
...
+module Main (main) where
+
+import Codec.Picture
+import Codec.Picture.Types -- to work with mutable images later
+import Control.Monad
+import Data.Array.Repa (Array, DIM1, DIM2, U, D, Z (..), (:.)(..), (!))
+import System.Environment (getArgs)
+import System.FilePath (replaceExtension)
+import qualified Data.Array.Repa as R -- for Repa
+```
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md:259]: @mrkkrp Would adding explicit imports help?

Also, would it make sense to add a type signature or something to make the exercise clearer?

File: tutorials/haskell/image-processing/tutorial.md [hunk 4]
```diff
@@ -0,0 +1,737 @@
...
+toExt Png      = "png"
+toExt Tiff     = "tiff"
+
+fromExt :: String -> Maybe ImgFormat
+fromExt "bmp"  = Just Bmp
+fromExt "jpeg" = Just Jpg
+fromExt "png"  = Just Png
+fromExt "tiff" = Just Tiff
+fromExt _      = Nothing
+```
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md:305]: @mrkkrp I haven't read the rest of the tutorial, but would splitting this and adding short explanations help? Or comments.

File: tutorials/haskell/image-processing/tutorial.md [hunk 5]
```diff
@@ -0,0 +1,737 @@
...
+
+As simple as that, we have rotated an image upside down:
+
+![Before rotation](/tutorials/haskell/image-processing/before-rotation.png)
+![After rotation](/tutorials/haskell/image-processing/after-rotation.png)
+
+Here is how our program performs:
+
+```
+$ image-processing before-rotation.png after-rotation.png +RTS -s
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md:403]: This shouldn't be a problem once we make the repository public so everyone can see the code, but we could think of saying that `image-processing` is the executable.

File: tutorials/haskell/image-processing/tutorial.md [hunk 6]
```diff
@@ -0,0 +1,737 @@
...
+
+withImage :: (Pixel a, PrimMonad m)
+  => Int                -- Image width
+  -> Int                -- Image height
+  -> (Int -> Int -> m a) -- Generating function
+  -> m (Image a)
+```
+
+If your image can be expressed as function from coordinates to pixels,
+`generateImage` looks super-simple and straightforward. `generateFildImage`
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: @mrkkrp `generateFoldImage`.

File: tutorials/haskell/image-processing/tutorial.md [hunk 7]
```diff
@@ -0,0 +1,737 @@
...
+allows to pass around an accumulator (i.e. you can have some sort of state).
+`withImage` is more interesting, it allows to use funtion that returns
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: @mrkkrp "is more interesting: it allows us to use a function"?

File: tutorials/haskell/image-processing/tutorial.md [hunk 8]
```diff
@@ -0,0 +1,737 @@
...
+
+The `extent` function returns shape of given array. We represent images as
+two-dimentional matrix of three-tuples holding components of pixel's color:
+`Array D DIM2 RGB8` and `Array U DIM2 RGB8` where `DIM2` is just built-in
+alias for `Z :. Int :. Int`.
+
+## Image rotation revisited
+
+Let's re-write code that rotates an image with Repa and see if it's worth
+it. To use Repa we need to compile our with the following flags:
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: @mrkkrp "our code"?

File: tutorials/haskell/image-processing/tutorial.md [hunk 9]
```diff
@@ -0,0 +1,737 @@
...
+[`repa-io`](https://hackage.haskell.org/package/repa-io) which can save in
+BMP (if that works for you), but for me it performed considerably worse than
+our `toImage` function in combination with `JuicyPixels`. There is also
+[`JuicyPixels-repa`](https://hackage.haskell.org/package/JuicyPixels-repa),
+but I got segfaults with it, no idea why.
+
+Please feel free to comment on the tutorial, I'm not an expert of Repa
+programming, in fact this is one of my first experiences with Repa, so maybe
+you can come up with idea how to improve the tutorial or maybe make the code
+faster. Thanks for reading!
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: Should we ask for feedback only once? We have a footer asking for feedback, but it doesn't have a lot of information, so we could write something better for that.

Comment (mrkkrp) [inline tutorials/haskell/image-processing/tutorial.md]: It may be better to ask for feedback “manually”, at least sometimes. People tend to pay more attention (if this is the right word) to custom, human-written content.

File: tutorials/haskell/image-processing/tutorial.md [hunk 10]
```diff
@@ -0,0 +1,743 @@
...
+author: Mark
+author-name: Mark Karpov
+---
+
+In this tutorial we will learn how to efficiently generate, transform, and
+process images in Haskell. This sort of task is interesting to solve in a
+pure functional language, because it's parallel in nature — processing of
+images requires performing identical manipulations on many pieces of data
+(pixels). On the other hand, such manipulation cannot be performed
+efficiently with help of familiar primitives for parallel Haskell computing:
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: @mrkkrp "the help". Perhaps adding parentheses helps:

> with the help of familiar primitives for parallel Haskell computing (the `Eval` monad, evaluation strategies, and the `Par` monad with `IVar`s), because...

Or dashes instead of parentheses. I added two more "the" above.

File: tutorials/haskell/image-processing/tutorial.md [hunk 11]
```diff
@@ -0,0 +1,743 @@
...
+`Eval` monad, evaluation strategies, and `Par` monad with `IVar`s, because
+these rely on laziness which works with boxed data, while for efficient
+processing of large arrays of numbers it's desirable to use unboxed data.
+
+This tutorial assumes basic knowledge of Haskell, as after reading
+[Learn You a Haskell for Great Good](http://learnyouahaskell.com/) or
+similar book. You don't need to know much more than that, but some
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "a similar book" or "similar books".

File: tutorials/haskell/image-processing/tutorial.md [hunk 12]
```diff
@@ -0,0 +1,743 @@
...
+unfamiliar terms will show up unavoidably, so if you don't yet understand
+underlying concepts, it's OK, you still can use the libraries, get the work
+done, and return to the subtle topics later (I give links for further
+reading in this article every time some mysterious Haskell concept shows
+up).
+
+## Available libraries
+
+There are a number of libraries for image processing available in Haskell
+ecosystem, some make use of external tools written in other languages,
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: @mrkkrp "the Haskell ecosystem". Instead of a comma, a colon or a dash would work better.

File: tutorials/haskell/image-processing/tutorial.md [hunk 13]
```diff
@@ -0,0 +1,743 @@
...
+  lightweight library written in Haskell that allows to load and save images
+  in various formats. It can be used to convert, generate, and process
+  images. Pixels are packed in unboxed storable vector — no automatic
+  parallel execution, but quite fast for simple tasks. We cover the library
+  in the tutorial.
+
+* [`repa`](https://hackage.haskell.org/package/repa) stands for “REgular
+  PArallel arrays” and is not specifically tied to image processing, but
+  it's the standard choice when you want to perform calculations on large
+  collection of numeric data in parallel in Haskell. We cover how to use it
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md:52]: "a large collection" or "large collections".

File: tutorials/haskell/image-processing/tutorial.md [hunk 14]
```diff
@@ -0,0 +1,743 @@
...
+  for image processing in this tutorial.
+
+* [`friday`](https://hackage.haskell.org/package/friday) is a powerful and
+  relatively new package that allows to manipulate images and has a good
+  collection of operations already coded for you (like edge detection,
+  cropping, filtering, etc.). Its principles of operation are very similar
+  to those of Repa, so after reading this tutorial you should have good
+  intuition how to code with `friday` too. To load images
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "intuition of how to" and maybe a comma after "To load images".

File: tutorials/haskell/image-processing/tutorial.md [hunk 15]
```diff
@@ -0,0 +1,743 @@
...
+  [`friday-devil`](https://hackage.haskell.org/package/friday-devil) is
+  usually used, but note that it requires external
+  [DevIL](http://openil.sourceforge.net/) library.
+
+* [`gloss`](https://hackage.haskell.org/package/gloss) describes itself as
+  “Painless 2D vector graphics, animations and simulations”. It indeed can
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: I think "Indeed, it can..." sounds better.

File: tutorials/haskell/image-processing/tutorial.md [hunk 16]
```diff
@@ -0,0 +1,743 @@
...
+  do more than just image editing and there are some nice demos to get you
+  started creating animations and simulations. Note that `gloss` uses
+  [OpenGL](https://www.opengl.org/) under the hood.
+
+We will use lightweight and “all in Haskell” `JuicyPixels` for reading and
+writing image files and `repa` for efficient processing of numeric data.
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: - "the lightweight".
- Do the quotes surrounding "all in Haskell" mean something in particular?
- We can add a comma before "and `repa`".

File: tutorials/haskell/image-processing/tutorial.md [hunk 17]
```diff
@@ -0,0 +1,743 @@
...
+
+## Juicy Pixels
+
+`JuicyPixels` is useful on its own if your task does not require parallel
+computations. Simple things like badge or identicon generation can be done
+sequentially (overhead of scheduling parallel execution for small amounts of
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "the overhead". Below, I think "the JuicyPixels API" (in this case, without the "'") sounds better.

File: tutorials/haskell/image-processing/tutorial.md [hunk 18]
```diff
@@ -0,0 +1,743 @@
...
+data can even make sequential execution preferable) and `JuicyPixels`' API
+makes them ridiculously simple.
+
+### Types and data structures used in Juicy Pixels
+
+If we quickly glance through
+[`Codec.Picture`](https://hackage.haskell.org/package/JuicyPixels/docs/Codec-Picture.html)
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "the Codec.Picture" module.

File: tutorials/haskell/image-processing/tutorial.md [hunk 19]
```diff
@@ -0,0 +1,743 @@
...
+      -- Internally pixel data is stored as consecutively packed
+      -- lines from top to bottom, scanned from left to right
+      -- within individual lines, from first to last color
+      -- component within each pixel.
+    , imageData   :: V.Vector (PixelBaseComponent a)
+    }
+```
+
+The definition is simple and intuitive: we have width and height of image
+and single vector that contains pixel data.
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "we have the width and height of an image, and a single vector that...".

File: tutorials/haskell/image-processing/tutorial.md [hunk 20]
```diff
@@ -0,0 +1,743 @@
...
+
+`!` before `Int` is called “strictness annotation”. Haskell, being a lazy
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "The ! before Int is called a 'strictness annotation'" or "is a 'strictness annotation'".

File: tutorials/haskell/image-processing/tutorial.md [hunk 21]
```diff
@@ -0,0 +1,743 @@
...
+language (should I have said “the lazy language”?) does not normally
+evaluate fields inside of data structure when data constructor itself is
+evaluated. When only constructor is evaluated data is said to be in “weak
+head normal form” for some obscure historical reason (fully evaluated data
+is said to be in “normal form”). The `!` sign says: “Hey, we are all in one
+boat now, if you evaluate constructor, evaluate me too”.
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: There's a comma missing before "does not normally..." and the rest can be as follows:

> does not normally evaluate fields inside of a data structure when the data constructor itself is evaluated. When only the constructor is evaluated, data is said to be in "weak head normal form"...

(There's one more "the" and a comma above.)

And "if you evaluate the constructor, evaluate me too" at the end.

File: tutorials/haskell/image-processing/tutorial.md [hunk 22]
```diff
@@ -0,0 +1,743 @@
...
+
+Here is how it works. Suppose we have this data type:
+
+```haskell
+data MyData = MyData !Int Int
+```
+
+One `Int` has strictness annotation while the other does not. Load the
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "a strictness annotation".

(I like this example with `:sprint`.)

File: tutorials/haskell/image-processing/tutorial.md [hunk 23]
```diff
@@ -0,0 +1,743 @@
...
+about.
+
+`{-# UNPACK #-}` is a different story. To make things lazy, Haskell by
+default represents everything as “boxed” data, that is, it stores a pointer
+to `Int` in `MyData`, not the `Int` itself:
+
+![Boxed data in Haskell](/tutorials/haskell/image-processing/boxed-data.png)
+
+[Thunk](https://wiki.haskell.org/Thunk) is a name for data that hasn't been
+evaluated yet. When `{-# UNPACK #-}` pragma is used, data becomes part of
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "When the `{-# UNPACK #-}` is used".

File: tutorials/haskell/image-processing/tutorial.md [hunk 24]
```diff
@@ -0,0 +1,743 @@
...
+parent structure:
+
+![Unboxed data in Haskell](/tutorials/haskell/image-processing/unboxed-data.png)
+
+To do something with `Int`s inside we don't need to de-reference pointer and
+that makes things faster. Of course in this case we cannot retain high level
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: Can you rephrase this part? I'm not sure if something like the following works:

> To do something with the `Int`s inside, we don't need to dereference the pointers, which makes things faster.

File: tutorials/haskell/image-processing/tutorial.md [hunk 25]
```diff
@@ -0,0 +1,743 @@
...
+of granularity with respect to lazy evaluation and we have to make unboxed
+data strict. Interestingly, GHC won't let you use `{-# UNPACK #-}` without
+strictness annotation saying that it's a parse error.
+
+I explained this in details because `imageData` is `Vector` of pixels and
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "in detail".

Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "a Vector of pixels".

File: tutorials/haskell/image-processing/tutorial.md [hunk 26]
```diff
@@ -0,0 +1,743 @@
...
+[`Par`](http://chimera.labs.oreilly.com/books/1230000000929/ch04.html) to
+compute different parts of it in parallel. Also, the `Vector` of pixels is
+immutable so every time you change it, you create a new vector copying all
+the data because normal Haskell's mechanisms for sharing data work well only
+with boxed data. We will return to this when we start talking about Repa and
+will see how the problem is solved there.
+
+To give the programmer single image type that represents “just” an image,
+abstracted from its representation, `Image` is put into `DynamicImage`
+wrapper:
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md:175]: > To give the programmer a single image type that represents 'just' an image abstracted from its representation, `Image` is put into a/the `DynamicImage` wrapper:

File: tutorials/haskell/image-processing/tutorial.md [hunk 27]
```diff
@@ -0,0 +1,743 @@
...
+     | ImageYCbCr8 (Image PixelYCbCr8)
+       -- | An image in the colorspace CMYK
+     | ImageCMYK8  (Image PixelCMYK8)
+       -- | An image in the colorspace CMYK and 16 bits precision
+     | ImageCMYK16 (Image PixelCMYK16)
+```
+
+Its data constructors are enumeration of images that correspond to supported
+pixel types. The result is that we can use `DynamicImage` type without
+caring much about concrete underlying representation.
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "an enumeration", "the `DynamicImage` type", and "the concrete underlying representation".

File: tutorials/haskell/image-processing/tutorial.md [hunk 28]
```diff
@@ -0,0 +1,743 @@
...
+
+Pixels in turn are just collection of numeric components, for example:
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "a collection".

File: tutorials/haskell/image-processing/tutorial.md [hunk 29]
```diff
@@ -0,0 +1,743 @@
...
+
+`saveJpgImage` takes an integer specifying desired quality as well. It's
+time to use this straightforward API for real work.
+
+### Format conversion program
+
+We can put together an utility for image format conversion right now! Let's
+write a console program that takes format of resulting image as first
+argument and path to source image as second argument. When the program is
+run, it converts image and saves it changing its extension appropriately.
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: > Let's write a console program that takes the format of the resulting image as a/the first argument and the path to the source image as a/the second argument. (...) it converts the image and...

I'm not sure about the a/the before first/second argument.

File: tutorials/haskell/image-processing/tutorial.md [hunk 30]
```diff
@@ -0,0 +1,743 @@
...
+
+Try to do it yourself first, here are the imports and language extensions
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: Instead of a comma, a dot or a dash would work better.

File: tutorials/haskell/image-processing/tutorial.md (hunk 12 above)
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: Not colon. Dot.

File: tutorials/haskell/image-processing/tutorial.md [hunk 31]
```diff
@@ -0,0 +1,743 @@
...
+fromExt "tiff" = Just Tiff
+fromExt _      = Nothing
+```
+
+We just built something useful with almost no effort, but what are our
+options if we wish to edit or generate an image?
+
+### Image rotation
+
+For processing of existing images the `pixelMap` function can be used:
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: > For processing existing images, the `pixelMap`...

File: tutorials/haskell/image-processing/tutorial.md [hunk 32]
```diff
@@ -0,0 +1,743 @@
...
+```
+
+With this function you can do something funny with colors, but let it be an
+exercise for the reader. How about operations that require random access to
+image's data? An example of such an operation is rotation.
+
+Since type at the base of `Image` is storable `Vector` (immutable, unboxed
+vector) and standard technique for efficient vector generation is via use of
+its mutable variant, it's no surprise that `JuicyPixels` provides
+`MutableImage` type.
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md:325]: How about something like the following?

> Since the type at the base of `Image` is a storable `Vector` (...) and the standard technique for efficient vector generation is using a mutable variant, it's no...

File: tutorials/haskell/image-processing/tutorial.md [hunk 33]
```diff
@@ -0,0 +1,743 @@
...
+
+Basic workflow is the same as with vectors:
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "The basic workflow".

File: tutorials/haskell/image-processing/tutorial.md [hunk 34]
```diff
@@ -0,0 +1,743 @@
...
+
+1. Allocate memory for new mutable image (or create it from immutable one).
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "Allocate memory for a new mutable image (or create it from an immutable one)".

File: tutorials/haskell/image-processing/tutorial.md [hunk 35]
```diff
@@ -0,0 +1,743 @@
...
+
+-- | Yield a mutable copy of an image by making a copy of it.
+thawImage :: (Storable (PixelBaseComponent px), PrimMonad m)
+  => Image px
+  -> m (MutableImage (PrimState m) px)
+```
+
+Reading and writing are done via `readPixel` and `writePixel` (they have
+unsafe companions but we won't touch them here). `freezeImage` helps with
+freezing. Let put it all together:
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "Let's".

File: tutorials/haskell/image-processing/tutorial.md [hunk 36]
```diff
@@ -0,0 +1,743 @@
...
+  freezeImage mimg
+```
+
+As simple as that, we have rotated an image upside down:
+
+![Before rotation](/tutorials/haskell/image-processing/before-rotation.png)
+![After rotation](/tutorials/haskell/image-processing/after-rotation.png)
+
+Here is how our program performs (`image-processing` is name of compiled
+executable):
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md:400]: "`image-processing` is the name of the compiled executable".

File: tutorials/haskell/image-processing/tutorial.md [hunk 37]
```diff
@@ -0,0 +1,743 @@
...
+  -> (acc, Image a)
+
+withImage :: (Pixel a, PrimMonad m)
+  => Int                -- ^ Image width
+  -> Int                -- ^ Image height
+  -> (Int -> Int -> m a) -- ^ Generating function
+  -> m (Image a)
+```
+
+If your image can be expressed as function from coordinates to pixels,
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "as a function".

File: tutorials/haskell/image-processing/tutorial.md [hunk 38]
```diff
@@ -0,0 +1,743 @@
...
+`generateImage` looks super-simple and straightforward. `generateFoldImage`
+allows to pass around an accumulator (i.e. you can have some sort of state).
+`withImage` is more interesting: it allows to use funtion that returns
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: s/funtion/function.

File: tutorials/haskell/image-processing/tutorial.md [hunk 39]
```diff
@@ -0,0 +1,743 @@
...
+values inside instances of `PrimMonad`. `PrimMonad` is outside of scope of
+this tutorial, but if you are interested you can read about it
+[here](https://www.schoolofhaskell.com/user/commercial/content/primitive-haskell).
+Basically, `PrimMonad` means `IO` or `ST` and monad stacks with one of these
+monads at the bottom.
+
+Let generate an image using `generateImage` function and save it as a PNG
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "Let's generate an image using the `generateImage` function...".

File: tutorials/haskell/image-processing/tutorial.md [hunk 40]
```diff
@@ -0,0 +1,743 @@
...
+
+```
+$ image-processing my-image.png +RTS -s
+…
+  Total   time    0.197s  (  0.104s elapsed)
+…
+```
+
+Generation of image is not parallel, the library just repeatedly calls
+provided function and builds resulting image from returned results. The
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "the library just repeatedly calls the provided function and build the resulting image...".

File: tutorials/haskell/image-processing/tutorial.md [hunk 41]
```diff
@@ -0,0 +1,743 @@
...
+performance is not bad at all, but can we do better using multiple cores?
+
+## Repa
+
+Here's where Repa comes into play. Operations on arrays performed in Repa
+are parallel by default if you compile your program with multi-thread
+support `-threaded` and run it with `+RTS -N`.
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: `-threaded` can be inside parentheses or something.

File: tutorials/haskell/image-processing/tutorial.md [hunk 42]
```diff
@@ -0,0 +1,743 @@
...
+
+Do you remember the problem with large unboxed arrays? We certainly cannot
+afford creation of new big immutable array after each operation, so how can
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: > We certainly cannot afford the creation of a new big immutable array after each...

File: tutorials/haskell/image-processing/tutorial.md [hunk 43]
```diff
@@ -0,0 +1,743 @@
...
+we avoid producing intermediate results?
+
+Well, the answer is rather simple. Following the same approach with function
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "a function".

File: tutorials/haskell/image-processing/tutorial.md [hunk 44]
```diff
@@ -0,0 +1,743 @@
...
+that takes coordinates as argument and returns a pixel, we can just make
+that single function more complex with help of primitives from functional
+programming like `map` but specialized for our purposes. Then we can call it
+to build the final result never creating intermediate ones.
+
+The final “real” array is called “manifest array” in Repa's terminology,
+while function that can be used to generate such real array is called
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "while the function that...". Below, "the friday package" and "the more mundane...".

File: tutorials/haskell/image-processing/tutorial.md [hunk 45]
```diff
@@ -0,0 +1,743 @@
...
+“delayed array”. The trick of never building intermediate results is called
+“fusion”. This is a popular technique that is also used in `friday` package
+and in more mundane `text`.
+
+### Repa Arrays
+
+Before we can start hacking with Repa, there is one more concept that needs
+to be explained: `Array` type and its shape. `Array r sh e` type is
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "the `Array` type".

File: tutorials/haskell/image-processing/tutorial.md [hunk 46]
```diff
@@ -0,0 +1,743 @@
...
+  [a few options](https://hackage.haskell.org/package/repa-3.4.0.2/docs/Data-Array-Repa.html)
+  available for that, but in this tutorial we will work only with `U`
+  (unboxed vector, manifest array), and `D` (delayed array, function from
+  indices to elements);
+
+* shape `sh`, this describes how many dimensions your array has and sizes of
+  these dimensions, below I show how to work with shapes;
+
+* element type `e`, in our case it will be `(Pixel8, Pixel8, Pixel8)` where
+  `Pixel8` is just a synonym for `Word8`, i.e. a byte.
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md:519]: We can change the first comma in each item for a colon. In the second item, a dot can be used instead of a comma before "below I show".

File: tutorials/haskell/image-processing/tutorial.md [hunk 47]
```diff
@@ -0,0 +1,743 @@
...
+```
+
+Take a look at Simon Marlow's
+[explanation of Repa shapes and indices](http://chimera.labs.oreilly.com/books/1230000000929/ch05.html#sec_par-repa-arrays)
+for more information.
+
+### Bridge from Juicy Pixels to Repa
+
+I will use two simple functions to convert between Juicy Pixels
+representation of pixels and Repa arrays, here they are:
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: Use a dot before "here they are" or just a colon and no "here they are".

File: tutorials/haskell/image-processing/tutorial.md [hunk 48]
```diff
@@ -0,0 +1,743 @@
...
+toImage a = generateImage gen width height
+  where
+    Z :. width :. height = R.extent a
+    gen x y =
+      let (r,g,b) = a ! (Z :. x :. y)
+      in PixelRGB8 r g b
+```
+
+The `extent` function returns shape of given array. We represent images as
+two-dimentional matrix of three-tuples holding components of pixel's color:
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "two-dimensional". Before that, "the shape".

File: tutorials/haskell/image-processing/tutorial.md [hunk 49]
```diff
@@ -0,0 +1,743 @@
...
+`Array D DIM2 RGB8` and `Array U DIM2 RGB8` where `DIM2` is just built-in
+alias for `Z :. Int :. Int`.
+
+## Image rotation revisited
+
+Let's re-write code that rotates an image with Repa and see if it's worth
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "the code".

File: tutorials/haskell/image-processing/tutorial.md [hunk 50]
```diff
@@ -0,0 +1,743 @@
...
+
+rotateImgRepa :: R.Source r e => Array r DIM2 e -> Array D DIM2 e
+rotateImgRepa g = R.backpermute e remap g
+  where
+    e@(Z :. width :. height) = R.extent g
+    remap (Z :. x :. y) = Z :. width - x - 1 :. height - y - 1
+    {-# INLINE remap #-}
+```
+
+`computeUnboxedP` builds manifest unboxed array in parallel. It's a
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "arrays in parallel" and "the more general" (below).

File: tutorials/haskell/image-processing/tutorial.md [hunk 51]
```diff
@@ -0,0 +1,743 @@
...
+type-specialized version of more general `computeP` function:
+
+```haskell
+computeP :: (Load r1 sh e, Target r2 e, Source r2 e, Monad m)
+  => Array r1 sh e     -- ^ The delayed array to compute
+  -> m (Array r2 sh e) -- ^ Manifest array — result
+```
+
+An interesting thing here is that `computeP` wants to live in monad, *any
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "in a monad". Below, "can use an array that...", "a run-time warning", a comma before "you are safe". At the end of the paragraph, "which may be a better choice".

File: tutorials/haskell/image-processing/tutorial.md [hunk 52]
```diff
@@ -0,0 +1,743 @@
...
+monad*, why is that? The reason is that `computeP` can use array that is
+produced with another `computeP` but only when it's already evaluated or
+else you get run-time warning and slow code. If you keep all `computeP` and
+`computeUnboxedP` functions in the same monad you are safe. So the type just
+helps avoid writing incorrect code (although it's still possible to use
+something like `runIdentity . computeP`). For sequential evaluation we have
+`computeS`, it may be a better choice for relatively small data-sets.
+
+`backpermute` is “backwards permutation of an array's elements” and it's
+just right tool for the job. Let's see what we get:
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "the right tool".

File: tutorials/haskell/image-processing/tutorial.md [hunk 53]
```diff
@@ -0,0 +1,743 @@
...
+
+```
+$ image-processing -- before-rotation.png after-rotation.png +RTS -s -N2
+…
+  Total   time    0.271s  (  0.156s elapsed)
+…
+```
+
+This seems to perform a bit better, although both Juicy Pixel's solution and
+Repa solution have varying time of execution every time you run them. We
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "both JuicyPixel's and Repa's solutions have...".

File: tutorials/haskell/image-processing/tutorial.md [hunk 54]
```diff
@@ -0,0 +1,743 @@
...
+```
+$ image-processing my-image.png +RTS -s -N4
+…
+  Total   time    0.152s  (  0.084s elapsed)
+…
+```
+
+### Conclusion
+
+Repa is great but watch out where is your bottleneck. If you're doing a lot
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "Repa is great but watch out for where your bottleneck is"?

File: tutorials/haskell/image-processing/tutorial.md [hunk 55]
```diff
@@ -360,21 +361,34 @@ main = do
...
+              (imageHeight - y - 1)
+              (pixelAt img x y)
+            go (x + 1) y
+  go 0 0
+```
+
+In `rotateImg`, we run a stateful computation inside the `ST` monad. You can
+think about `ST` monad as an “escapable” `IO`. References to mutable data
+cannot escape the `ST` monad, but pure data can. `ST` monad is often used to
+write functions that manipulate mutable values as long as result is
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "The ST monad" in line 385, and "as long as the result" in line 386.

File: tutorials/haskell/image-processing/tutorial.md [hunk 56]
```diff
@@ -360,21 +361,34 @@ main = do
...
+referentially transparent.
+
+We use `unsafeFreezeImage` instead of `freezeImage` because it allows us
+avoid unnecessary copying of data. Since after “freezing” we don't modify
```
Comment (jpvillaisaza) [inline tutorials/haskell/image-processing/tutorial.md]: "because it allows us to avoid unnecessary copying of data" or:

> We use `unsafeFreezeImage` instead of `freezeImage` to avoid unnecessary copying of data.

**Comment by jpvillaisaza** [comment]
@mrkkrp A couple more comments:
- In the first paragraph of the JuicyPixels section, change "and JuicyPixels API makes them ridiculously simple" for "and **the** JuicyPixels API makes them ridiculously simple".
- There are some links to specific code in the JuicyPixels library. The links in the tutorial don't include a specific version, so they take you to the latest version (3.2.7.1), but the Tested with section says version 3.2.7. Can you add the version number to the appropriate links? I think this is only a problem for JuicyPixels, not Repa.
- In the Repa arrays subsection, change "one-dimentional" for "one-dimensional".

After applying these changes, please change the date of the tutorial to June 6 (2016-06-06) and merge to the tutorials branch.

Thanks!
